fix: Correctly cast from UTF16 positions

tombh opened the pull request:
Fixes #302 

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

pyscripter commented on pygls/workspace.py:
```diff
@@ -80,14 +84,42 @@ def position_from_utf16(lines: List[str], position: Position) -> Position:
+        )
+        log.debug(f"position_from_utf16() line with {utf16_num_units(line)} UTF16 units: `{line}`")
+        position.character = utf16_num_units(line) - 1
+
```
position can be greater than utf16_num_units(line), if for example the cursor is at the end of the line.
I  would just drop this whole block not least because utf16_num_units slows things down.
_apply_incremental_change is designed to work correctly with positions beyond the end of the line.

tombh replied:
i've kept this just to satisfy this test `assert doc.offset_at_position(Position(line=1, character=5)) == 12`, which seems to make sense to me

tombh replied:
@pyscripter I need some help to understand why you consider that existing test line to be wrong. Do you disagree with this assertion `assert doc.offset_at_position(Position(line=1, character=5)) == 12`?

pyscripter commented on pygls/workspace.py:
```diff
@@ -80,14 +84,42 @@ def position_from_utf16(lines: List[str], position: Position) -> Position:
         )
-    except IndexError:
-        return Position(line=len(lines), character=0)
+        line = lines[-1]
```
There is no need to issue a waring.
_apply_incremental_change is designed to work correctly with positions beyond the end of the file.

I  would just replace all the above with:
```python
    if position.line >= len(lines):
        return position

    line = lines[position.line]
```

pyscripter commented on pygls/workspace.py:
```diff
@@ -80,14 +84,42 @@ def position_from_utf16(lines: List[str], position: Position) -> Position:
+    _utf16_index = 0
+    _is_end_of_line = False
+    while True:
+        _current_char = line[conventional_index]
```
This would fail on empty lines

pyscripter commented:
Please see the comments in the code.   I would just use (skipping the comments):

```python
def position_from_utf16(lines: List[str], position: Position) -> Position:
    if position.line >= len(lines):
        # start of the line after last
        return Position(len(lines), 0)  # or return position

    line = lines[position.line]

    _utf32_len = len(line)
    _utf32_index = 0
    _utf16_index = 0
    while (_utf16_index < position.character) and (_utf32_index < _utf32_len):
        _current_char = line[line_index]
        is_double_width = is_char_beyond_multilingual_plane(_current_char)
        if (is_double_width):
            _utf16_index += 2
        else:
            _utf16_index += 1
        _utf32_index += 1

    position = Position(
        line=position.line,
        character=_utf32_index
    )
    return position

```

tombh commented:
Thanks for the feedback and extra test cases, they're really useful. Before making these amends, can you have a look at Pygls' [`Document.offset_at_position()`](https://github.com/openlawlibrary/pygls/blob/2a052539aef0e5ada2c619ae7df9baae5e7f8939/tests/test_document.py#L250). It's the main reason I made those conditional guards at the beginning of the function. I don't know who uses `offset_at_position()`, or what it's for, but Pygls supports it, so we should probably continue to do so.

pyscripter commented:
> I don't know who uses `offset_at_position()`

Well, some editors may internally store the actual text of a file instead of a list of lines like pygls does and they may wont to convert from a position to an offset to the string representation of the file.  Since offset_at_position does not account for line breaks, users of the function need to adjust for line breaks.

The function will work correctly if position_from_utf16 provides the correct result, so it should not be a concern.

tombh commented:
Ok I've pushed some changes. I made the variables names a bit more verbose. I added a guard clause for when a queried `Position` goes over the length of the line. And I added all your tests cases.

How does it look to you?

tombh commented:
I'm thinking of merging this as is...

pyscripter commented:
_utf16_end_of_line = utf16_num_units(_line) - 1

This is definitely wrong.  See my comments above.

tombh commented:
Did you miss my question here? https://github.com/openlawlibrary/pygls/pull/304#discussion_r1089369599

tombh commented:
Just taking a small step back for a moment. This is surprisingly a lot more complicated than I first thought. We have 2 things going on here.

1. The first one, and the reason that this PR exists, is the `position_from_utf16` bug you identified in #302. And I think the PR here addresses that?
2. The second one is the tangentially associated `offset_at_position` tests that fail because of the fix for 1. (We have some comments on that in [a now stale commit](https://github.com/openlawlibrary/pygls/commit/3b36463bf1476e51758a04dfee2fc6024b388001).)

Copying your comments from point 2 here:
> But why include the control line end characters in lines?
> ...
> You also seem to assume that the control characters are always /n. This is not true on Windows or for the last line.

Very good questions. I had to look this up and you were indeed right to bring it up. If I'm interpreting it correctly the LSP spec says that that `\n` and `\r\n` should be considered and considered as the same length. [The relevant section is here](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments). And I've added a new test for that called `test_position_for_line_endings`.

To further complicate things, I also learnt from reading the spec that the position encodings can be negotiated by the client and server! See:
![image](https://github.com/openlawlibrary/pygls/assets/160835/b71888e1-1e13-4c60-af3c-cfa02357df09)
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#clientCapabilities

It does say that it defaults to utf-16 positioning, which is what Pygls does already. So maybe we can just create an issue to acknowledge that Pygls doesn't currently support clients requesting specific position encodings.

tombh commented:
I'm thinking of merging this then.

alcarney reviewed (approved):
I've never managed to get my head around this issue, so if you say it's ready I'll take your word for it! :smile:

tombh commented:
After 8 months, it's as ready as it'll ever be I think!
